### PR TITLE
charmstore: add CORS headers to all requests

### PR DESCRIPTION
**Pull request by rogpeppe**
To QA this, we need to create a web page that contains some javascript
that makes an AJAX request to the charm store, and check that it
succeeds.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/497/

File: internal/router/router.go [hunk 1]
```diff
@@ -152,6 +152,27 @@ func New(handlers *Handlers, resolveURL func(url *charm.Reference) error) *Route
...
 // ServeHTTP implements http.Handler.ServeHTTP.
 func (r *Router) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	// Allow cross-domain access from anywhere, including AJAX
+	// requests. An AJAX request will add an X-Requested-With:
+	// XMLHttpRequest header, which is a non-standard header, and
+	// hence will require a pre-flight request, so we need to
+	// specify that that header is allowed, and we also need to
+	// implement the OPTIONS method so that the pre-flight request
+	// can work.
+	// See https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS
```
Comment (frankban) [inline internal/router/router.go:162]: Thank you!

**Comment by urosj** [comment]
:+1: 
Thanks for the excellent documentation.

**Comment by frankban** [comment]
Nice branch with a simple solution.
:+1:

**Comment by jrwren** [comment]
:+1: Thanks for cleaning up the previous hack.

**Comment by rogpeppe** [comment]
QA'd by Makyo.

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/498/

**Pull request merged**